### PR TITLE
feat(core): add substream operator

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Fixes: N/A

## What is the new behavior?
Adds the substream operator for listening for a sequence of multiple actions.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

**Comment by lderrickable** [comment]
This looks really good to me. It's actually a bit different than how I was imagining it, but I don't think it could've been a pure function the way I was picturing it. Nice work!

File: libs/core/src/operators/substream.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+export type ActionType = Action['type'];
+export type ActionSequenceStep = ActionType | ActionType[];
+export type ActionSequence = ActionSequenceStep[];
+
+const compareActionSequenceStep = (step: ActionSequenceStep, action: Action): boolean => Array.isArray(step)
+  ? step.reduce((acc, type) => acc || action.type === type, false)
+  : action.type === step
+
+/**
+ * Finds and emits a substream of actions.
```
Comment (lderrickable) [inline libs/core/src/operators/substream.pipe.ts:64]: I think this should be something like: "Watches for a particular stream of actions, and emits a list of those actions once all actions have occurred"

File: libs/core/src/operators/substream.pipe.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * You can optionally pass a list of terminators.
+ * If a terminator is encountered in the action stream, any partially matched substream is reset.
+ * For example, if substream([A, B], C), an action stream of [A1, C1, B1] would cause substream to not emit.
+ *
+ * @param sequence The sequence of action types that define the substream to listen for.
+ * @param terminators A list of terminators, any of which interrupt and reset a partially matched substream.
+ */
+export const substream = (sequence: ActionSequence, ...terminators: ActionType[]) => {
+  const buildMatchedSequence = (matchedSequence, action) =>
```
Comment (lderrickable) [inline libs/core/src/operators/substream.pipe.ts:84]: I think this function should be named something else like "accumulateStream" or "accumulateSequence" or something like that.

Comment (griest024) [reply libs/core/src/operators/substream.pipe.ts:84]: How about `accumulateMatchedStream`? yeah, you're right what I call the sequence and the stream are different data structures. I shouldn't mix the terms.

File: libs/core/src/operators/substream.pipe.ts (hunk 1 above)
Comment (griest024) [reply libs/core/src/operators/substream.pipe.ts:64]: That's better. I think `Watches for a particular stream of actions in the specified order and emits a list of those actions once all sequence steps have been matched.` is even a little more precise.

File: libs/core/src/operators/substream.pipe.ts (hunk 2 above)
Comment (lderrickable) [reply libs/core/src/operators/substream.pipe.ts:84]: Oh, good point. I was actually just talking about the word "build". `accumulateMatchedSubstream` sounds good

Comment (griest024) [reply libs/core/src/operators/substream.pipe.ts]: oh right I don't think I read your comment fully. not sure what I was thinking. changed.

**Comment by griest024** [comment]
@lderrickable don't merge yet

**Comment by griest024** [comment]
@damienwebdev `bufferToggle` doesn't work that well. I vote either using this current implementation or  writing a better `buffer` operator.

**Comment by griest024** [comment]
@damienwebdev @lderrickable I have removed comparator support and added Damien's fix for not using `skipWhile` (I'm dumb).

**Pull request merged**